src/inmanta/server/services/orchestrationservice.py

```diff
@@ -63,277 +63,204 @@
-        :param partial_updates: The resources part of the partial compile
-        :param resource_states: The resource states of the resources in the partial compile.
-        :param resource_sets: The resource sets of the resources in the partial compile.
-        :param removed_resource_sets: The resources in these resource sets should be present in the new configuration model.
-        :param unknowns: The unknowns that belong to the partial model.
+    def _merge_shared_resources(self, shared_resources_new: Dict[ResourceIdStr, data.Resource]) -> abc.Sequence[data.Resource]:
         """
-        self.partial_updates = partial_updates
-        self.resource_states = resource_states
-        self.resource_sets = resource_sets
-        self.removed_resource_sets = removed_resource_sets
-        self.base_version: int = base_version
-        self.new_version: int = new_version
-        self.env = env
-        self.unknowns = unknowns
-
-    def pair_resources_partial_update_to_old_version(
-        self, old_resources: Dict[ResourceIdStr, ResourceWithResourceSet]
-    ) -> List[PairedResource]:
-        """
-        returns a list of paired resources
+        Merge the set of shared resources present in the old version of the model together with the set of shared resources
+        present in the partial compile.
 
-        :param old_resources: A dict with as Key an ResourceIdStr and as value a resource and its resource_set.
+        :param shared_resources_new: The set of shared resources present in the partial compile.
+        :returns: The set of shared resources that should be present in the new version of the model.
         """
-        paired_resources: List[PairedResource] = []
-        for partial_update in self.partial_updates:
-            key = Id.parse_id(partial_update.id).resource_str()
-            resource_set = self.resource_sets.get(key)
-            resource_state = self.resource_states.get(key, const.ResourceState.available)
-            pair: PairedResource = PairedResource(
-                new_resource=ResourceWithResourceSet(partial_update, resource_set, resource_state),
-                old_resource=old_resources.get(key),
-            )
-            paired_resources.append(pair)
-        return paired_resources
+        all_rids_shared_resources = set(self.shared_resources_old.keys()) | set(shared_resources_new.keys())
+        result = []
+        for rid_shared_resource in all_rids_shared_resources:
+            if rid_shared_resource in shared_resources_new and rid_shared_resource in self.shared_resources_old:
+                # Merge requires/provides shared resource
+                old_shared_resource = self.shared_resources_old[rid_shared_resource]
+                new_shared_resource = shared_resources_new[rid_shared_resource]
+                res = self._merge_requires_and_provides_of_shared_resource(old_shared_resource, new_shared_resource)
```

Comment:
it does provide some powerful capabilities!

Reply:
I'm convinced now! I do think a picture of resource sets and their allowed edges to the shared resource set (both for normal relations and requires-provides) in the docs would be very useful. Not necessarily as part of this ticket, since nothing new was introduced really, just refined.